synapse/storage/databases/main/event_push_actions.py

```diff
@@ -1154,6 +1275,64 @@ def _handle_new_receipts_for_notifs_txn(self, txn: LoggingTransaction) -> bool:
                 value_values=[(row[0], row[1]) for row in unread_counts],
             )
 
+        # For each new read receipt we delete push actions from before it and
+        # recalculate the summary.
+        for _, room_id, user_id, thread_id, stream_ordering in rows:
+            # Only handle our own read receipts.
+            if not self.hs.is_mine_id(user_id):
+                continue
+
+            if thread_id is None:
+                continue
+
+            # For each new read receipt we delete push actions from before it and
+            # recalculate the summary.
+            txn.execute(
+                """
+                DELETE FROM event_push_actions
+                WHERE room_id = ?
+                    AND user_id = ?
+                    AND thread_id = ?
+                    AND stream_ordering <= ?
+                    AND highlight = 0
+                """,
+                (room_id, user_id, thread_id, stream_ordering),
+            )
+
+            # Fetch the notification counts between the stream ordering of the
+            # latest receipt and what was previously summarised.
+            unread_counts = self._get_notif_unread_count_for_user_room(
+                txn,
+                room_id,
+                user_id,
+                stream_ordering,
+                old_rotate_stream_ordering,
+                thread_id,
+            )
+            # unread_counts will be a list of 0 or 1 items.
```

Comment:
We should assert this.

Also how do we know that it will be 0 or 1 items?

Reply:
> Also how do we know that it will be 0 or 1 items?

Because it is selecting on the given `thread_id`, so either that thread ID has new counts or it does not.

Reply:
It looks like this code was updated to handle all of the `unread_counts`, not just the first one 👍

https://github.com/matrix-org/synapse/blob/e7b5421e47c174b139da2eef3e032d7a40f93cb2/synapse/storage/databases/main/event_push_actions.py#L1294-L1310

Reply:
The code was combined, but for a thread there will only be 1 or 0.